File: go/internal/sol/p0089/p0089.go
package p0089

import (
	_ "embed"
	"regexp"
	"strconv"
	"strings"
)

//go:embed 0089_roman.txt
var fileContent string

func replaceRomanNumbers(line string) string {
	var s string

	// step 1a: IX/XC/CM
	s = regexp.MustCompile(`IIIIIIIII|XXXXXXXXX|CCCCCCCCC`).ReplaceAllString(line, "##")
	// step 1b: IX/XC/CM
	s = regexp.MustCompile(`VIIII|LXXXX|DCCCC`).ReplaceAllString(s, "##")
	// step 2: V/L/D
	s = regexp.MustCompile(`IIIII|XXXXX|CCCCC`).ReplaceAllString(s, "#")
	// step 3: IV/XL/CD
	s = regexp.MustCompile(`IIII|XXXX|CCCC`).ReplaceAllString(s, "##")

	return s
}

func compute(data string) string {
	var result int
	for _, line := range strings.Split(strings.Trim(data, "\n"), "\n") {
		result += len(line) - len(replaceRomanNumbers(line))
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute(fileContent)
}

File: haskell/src/Sol/P0089.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module Sol.P0089 (compute, solve) where

import qualified Data.ByteString.Char8 as BS (ByteString, lines)
import qualified Data.FileEmbed as FE (embedFile, makeRelativeToProject)
import qualified Data.Text as T (Text, length, replace)
import qualified Data.Text.Encoding as TE (decodeLatin1)

fileData :: BS.ByteString
fileData = $(FE.makeRelativeToProject "resources/0089_roman.txt" >>= FE.embedFile)

parseData :: BS.ByteString -> [T.Text]
parseData = map TE.decodeLatin1 . BS.lines

replaceNumStr :: T.Text -> T.Text
replaceNumStr =
    (flip . foldl)
        (\s (from, to) -> T.replace from to s)
        subsutitutionTbl
  where
    subsutitutionTbl :: [(T.Text, T.Text)]
    subsutitutionTbl =
        [ ("IIIIIIIII", "##")
        , ("XXXXXXXXX", "##")
        , ("CCCCCCCCC", "##")
        , ("VIIII", "##")
        , ("LXXXX", "##")
        , ("DCCCC", "##")
        , ("IIIII", "#")
        , ("XXXXX", "#")
        , ("CCCCC", "#")
        , ("IIII", "##")
        , ("XXXX", "##")
        , ("CCCC", "##")
        ]

compute :: String
compute =
    show . foldl (\acc s -> acc + T.length s - T.length (replaceNumStr s)) 0 $
        parseData fileData

solve :: String
solve = compute
